File: impl/csharp/test/DescriptionTests.cs
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.Json;

using Naalp;
using Xunit;

namespace Sh.Bubblefish.Naalp.Tests
{
    /// <summary>
    /// C18 signed description / directory / import conformance for the C# SDK (design.md §21;
    /// R-DESC-1..8), graded against the shared independent corpus <c>vectors/description/cases.json</c>
    /// (NOT produced by this code): the Description body/head/id and each operation body; the Directory
    /// bodies (version A, the fork B, the length-fork, the different-version succession); the fork
    /// FIRST-DIFFERING position; the Import body/head/id and the bound foreign-id; and the closed
    /// naalp-description-format rejection of an unknown format.
    ///
    /// <para>The offline-verification, directory fork-proof, and confused-deputy import paths use REAL
    /// deterministic ML-DSA-65 (BouncyCastle, rnd=0) and are demonstrated in isolation — the corpus
    /// carries no signed vector (F2/F4, honest). Ported from impl/go/description; mirrors
    /// impl/python/naalp/description.py.</para>
    ///
    /// <para>Deviation from the Go reference (honest F4 note): Go's VerifyImport carries an
    /// ErrVerifierKeyMismatch guard because it takes BOTH an (alg, pubkey) pair AND a separate
    /// cose.Verifier, and must bind them before deriving the authority id. This C# port — like the
    /// Python/Kotlin ports and the sibling <see cref="Gateway.VerifyDecision"/> — verifies with a SINGLE
    /// (alg, pubkey) pair, so the authority id is ALWAYS derived from exactly the key that verified the
    /// signature; the mismatch that guard prevents is structurally impossible here. There is no
    /// VerifierKeyMismatch surface to port: it is absent because the vulnerability it guards cannot
    /// arise in this signature, NOT silently dropped.</para>
    /// </summary>
    public sealed class DescriptionTests
    {
        private const int Alg = Cose.ALG_MLDSA65;

        private static JsonElement Vector()
        {
            var d = new DirectoryInfo(AppContext.BaseDirectory);
            for (int i = 0; i < 12 && d != null; i++)
            {
                string p = Path.Combine(d.FullName, "vectors", "description", "cases.json");
                if (File.Exists(p))
                {
                    return JsonDocument.Parse(File.ReadAllText(p)).RootElement.Clone();
                }
                d = d.Parent;
            }
            throw new Xunit.Sdk.XunitException("vectors/description/cases.json not found from " + AppContext.BaseDirectory);
        }

        private static string Hex(byte[] b) => Convert.ToHexString(b).ToLowerInvariant();
        private static byte[] Hb(string s) => Convert.FromHexString(s);

        private static List<Description.Operation> OpsOf(JsonElement arr)
        {
            var ops = new List<Description.Operation>();
            foreach (JsonElement o in arr.EnumerateArray())
            {
                ops.Add(new Description.Operation(
                    o.GetProperty("name").GetString()!,
                    o.GetProperty("effect").GetInt64(),
                    o.GetProperty("requires_approval").GetInt64()));
            }
            return ops;
        }

        private static Description.Doc DocOf(JsonElement dv)
        {
            return new Description.Doc(Hb(dv.GetProperty("service_hex").GetString()!), OpsOf(dv.GetProperty("operations")));
        }

        private static Description.Directory DirOf(byte[] id, long version, JsonElement membersHex)
        {
            var members = new List<byte[]>();
            foreach (JsonElement m in membersHex.EnumerateArray())
            {
                members.Add(Hb(m.GetString()!));
            }
            return new Description.Directory(id, version, members);
        }

        // ---- Description: the signed operation table body + each operation body --------------------

        [Fact]
        public void DescriptionBodyMatchesOracle()
        {
            JsonElement v = Vector();
            JsonElement dv = v.GetProperty("description");
            Description.Doc doc = DocOf(dv);
            Assert.Equal(dv.GetProperty("body_hex").GetString(), Hex(doc.Bytes()));
            Assert.Equal(dv.GetProperty("head_hex").GetString(), Hex(doc.Head()));
            Assert.Equal(dv.GetProperty("id_hex").GetString(), Hex(doc.Id()));

            foreach (JsonElement o in dv.GetProperty("operations").EnumerateArray())
            {
                var op = new Description.Operation(
                    o.GetProperty("name").GetString()!, o.GetProperty("effect").GetInt64(), o.GetProperty("requires_approval").GetInt64());
                Assert.Equal(o.GetProperty("body_hex").GetString(), Hex(op.Bytes()));
            }

            // ParseDescription reconstructs the table from the bytes ALONE (offline-verifiable).
            Description.Doc parsed = Description.ParseDescription(Hb(dv.GetProperty("body_hex").GetString()!));
            Assert.Equal(dv.GetProperty("body_hex").GetString(), Hex(parsed.Bytes()));
            Assert.True(parsed.Operation("purge", out Description.Operation purge));
            Assert.Equal(Policy.DESTRUCTIVE, purge.EffectClass());
            Assert.True(purge.RequiresApprovalFlag());
        }

        // ---- Directory: the version A body, the fork B, the length-fork, the succession ------------

        [Fact]
        public void DirectoryBodiesMatchOracle()
        {
            JsonElement v = Vector();
            JsonElement dir = v.GetProperty("directory");
            byte[] id = Hb(dir.GetProperty("directory_hex").GetString()!);
            long ver = dir.GetProperty("version").GetInt64();

            Description.Directory a = DirOf(id, ver, dir.GetProperty("members_a_hex"));
            JsonElement av = dir.GetProperty("a");
            Assert.Equal(av.GetProperty("body_hex").GetString(), Hex(a.Bytes()));
            Assert.Equal(av.GetProperty("head_hex").GetString(), Hex(a.Head()));
            Assert.Equal(av.GetProperty("id_hex").GetString(), Hex(a.Id()));

            JsonElement forkV = dir.GetProperty("fork");
            Description.Directory b = DirOf(id, ver, forkV.GetProperty("members_b_hex"));
            JsonElement bv = forkV.GetProperty("b");
            Assert.Equal(bv.GetProperty("body_hex").GetString(), Hex(b.Bytes()));
            Assert.Equal(bv.GetProperty("head_hex").GetString(), Hex(b.Head()));
            Assert.Equal(bv.GetProperty("id_hex").GetString(), Hex(b.Id()));

            JsonElement lf = dir.GetProperty("length_fork");
            Description.Directory shortDir = DirOf(id, ver, lf.GetProperty("members_short_hex"));
            Assert.Equal(lf.GetProperty("body_hex").GetString(), Hex(shortDir.Bytes()));

            JsonElement dvv = dir.GetProperty("different_version");
            Description.Directory v8 = DirOf(id, dvv.GetProperty("version").GetInt64(), forkV.GetProperty("members_b_hex"));
            Assert.Equal(dvv.GetProperty("body_hex").GetString(), Hex(v8.Bytes()));
        }

        // ---- fork detection at the FIRST-DIFFERING member position ---------------------------------
        // MUTATION ANCHOR: dropping the element-comparison loop in FirstMemberDifference makes the
        // same-length fork (A vs B, member[1] differs) return fork=false, failing Assert.True(fork).

        [Fact]
        public void DirectoryForkAtPosition()
        {
            JsonElement v = Vector();
            JsonElement dir = v.GetProperty("directory");
            byte[] id = Hb(dir.GetProperty("directory_hex").GetString()!);
            long ver = dir.GetProperty("version").GetInt64();

            Description.Directory a = DirOf(id, ver, dir.GetProperty("members_a_hex"));
            Description.Directory b = DirOf(id, ver, dir.GetProperty("fork").GetProperty("members_b_hex"));
            Description.Directory shortDir = DirOf(id, ver, dir.GetProperty("length_fork").GetProperty("members_short_hex"));
            Description.Directory v8 = DirOf(id, dir.GetProperty("different_version").GetProperty("version").GetInt64(),
                dir.GetProperty("fork").GetProperty("members_b_hex"));

            (int pos, bool fork) = Description.DetectFork(a, b);
            Assert.True(fork);
            Assert.Equal(dir.GetProperty("fork").GetProperty("first_differing_position").GetInt32(), pos);

            (int lpos, bool lfork) = Description.DetectFork(a, shortDir);
            Assert.True(lfork);
            Assert.Equal(dir.GetProperty("length_fork").GetProperty("first_differing_position").GetInt32(), lpos);

            // identical members => benign duplicate, not a fork (the corpus -1 sentinel).
            (_, bool dupFork) = Description.DetectFork(a, a);
            Assert.False(dupFork);
            Assert.Equal(-1, dir.GetProperty("duplicate_first_differing_position").GetInt32()); // corpus sentinel

            // a different version is a legitimate succession, not a fork.
            (_, bool verFork) = Description.DetectFork(a, v8);
            Assert.False(verFork);

            // a different directory id is not a conflicting pair.
            Description.Directory other = DirOf(Hb("00"), ver, dir.GetProperty("members_a_hex"));
            (_, bool otherFork) = Description.DetectFork(a, other);
            Assert.False(otherFork);
        }

        // ---- Import: the attestation body + the bound foreign-id, and unknown-format rejection -----

        [Fact]
        public void ImportBodyMatchesOracle()
        {
            JsonElement v = Vector();
            JsonElement im = v.GetProperty("import");
            var imp = new Description.Import(
                Hb(im.GetProperty("importer_hex").GetString()!),
                im.GetProperty("format").GetInt64(),
                Hb(im.GetProperty("foreign_hex").GetString()!),
                OpsOf(im.GetProperty("operations")));

            Assert.Equal(im.GetProperty("body_hex").GetString(), Hex(imp.Bytes()));
            Assert.Equal(im.GetProperty("head_hex").GetString(), Hex(imp.Head()));
            Assert.Equal(im.GetProperty("id_hex").GetString(), Hex(imp.Id()));
            Assert.Equal(im.GetProperty("foreign_id_hex").GetString(), Hex(imp.ForeignId()));

            // ParseImport round-trips the body bytes.
            Description.Import parsed = Description.ParseImport(Hb(im.GetProperty("body_hex").GetString()!));
            Assert.Equal(im.GetProperty("body_hex").GetString(), Hex(parsed.Bytes()));

            // a format outside the closed naalp-description-format set {1,2,3} is rejected on decode.
            JsonElement uf = im.GetProperty("unknown_format");
            var ex = Assert.Throws<NaalpException>(() => Description.ParseImport(Hb(uf.GetProperty("body_hex").GetString()!)));
            Assert.Equal(uf.GetProperty("reject").GetString(), ex.Kind); // UnknownDescriptionFormat
        }

        // ---- offline verification + the non-repudiable directory fork-proof (REAL ML-DSA) ----------

        [Fact]
        public void OfflineVerifyAndForkProofInIsolation()
        {
            JsonElement v = Vector();
            JsonElement dir = v.GetProperty("directory");
            byte[] id = Hb(dir.GetProperty("directory_hex").GetString()!);
            long ver = dir.GetProperty("version").GetInt64();
            int wantPos = dir.GetProperty("fork").GetProperty("first_differing_position").GetInt32();

            byte[] seed = new byte[32];
            for (int i = 0; i < 32; i++) seed[i] = 21;
            byte[] pk = Cose.MldsaKeygen("ML-DSA-65", seed);
            string signerId = Identity.SignerId(Alg, pk);

            Description.Directory a = DirOf(id, ver, dir.GetProperty("members_a_hex"));
            Description.Directory b = DirOf(id, ver, dir.GetProperty("fork").GetProperty("members_b_hex"));
            Description.Directory v8 = DirOf(id, dir.GetProperty("different_version").GetProperty("version").GetInt64(),
                dir.GetProperty("fork").GetProperty("members_b_hex"));
            byte[] signedA = Description.SignDirectory(a, Alg, seed);
            byte[] signedB = Description.SignDirectory(b, Alg, seed);
            byte[] signedV8 = Description.SignDirectory(v8, Alg, seed);

            // offline verification: the SAME signed bytes reconstruct the directory regardless of host.
            Description.Directory ra = Description.VerifyDirectory(signedA, Cose.PROFILE_PUBLIC, Alg, pk);
            Assert.Equal(Hex(a.Bytes()), Hex(ra.Bytes()));

            // the honest fork proof: two signed directories at (dir, ver) with differing members.
            var fp = new Description.DirectoryForkProof(Encoding.UTF8.GetBytes(signerId), signedA, signedB);
            int pos = fp.Verify(Cose.PROFILE_PUBLIC, Alg, pk);
            Assert.Equal(wantPos, pos);

            // an unnamed accused is not evidence.
            var unnamed = new Description.DirectoryForkProof(Array.Empty<byte>(), signedA, signedB);
            var uEx = Assert.Throws<NaalpException>(() => unnamed.Verify(Cose.PROFILE_PUBLIC, Alg, pk));
            Assert.Equal("DirForkProofInvalid", uEx.Kind);

            // identical members are a benign duplicate, not a fork.
            var dup = new Description.DirectoryForkProof(Encoding.UTF8.GetBytes(signerId), signedA, signedA);
            var dEx = Assert.Throws<NaalpException>(() => dup.Verify(Cose.PROFILE_PUBLIC, Alg, pk));
            Assert.Equal("DirForkProofInvalid", dEx.Kind);

            // a different version is a succession, not a fork.
            var succ = new Description.DirectoryForkProof(Encoding.UTF8.GetBytes(signerId), signedA, signedV8);
            var sEx = Assert.Throws<NaalpException>(() => succ.Verify(Cose.PROFILE_PUBLIC, Alg, pk));
            Assert.Equal("DirForkProofInvalid", sEx.Kind);

            // a signature that does not verify under the key is rejected BadSignature.
            byte[] foreignSeed = new byte[32];
            for (int i = 0; i < 32; i++) foreignSeed[i] = 99;
            byte[] foreignPk = Cose.MldsaKeygen("ML-DSA-65", foreignSeed);
            var fEx = Assert.Throws<NaalpException>(() => fp.Verify(Cose.PROFILE_PUBLIC, Alg, foreignPk));
            Assert.Equal("BadSignature", fEx.Kind);
        }

        // ---- the confused-deputy rule: the importer is the sole authority (REAL ML-DSA) ------------

        [Fact]
        public void ImportConfusedDeputyInIsolation()
        {
            JsonElement v = Vector();
            JsonElement im = v.GetProperty("import");
            byte[] foreign = Hb(im.GetProperty("foreign_hex").GetString()!);
            List<Description.Operation> ops = OpsOf(im.GetProperty("operations"));

            byte[] seed = new byte[32];
            for (int i = 0; i < 32; i++) seed[i] = 33;
            byte[] pk = Cose.MldsaKeygen("ML-DSA-65", seed);
            string signerId = Identity.SignerId(Alg, pk);

            // the importer field MUST be the wrapping signer's own self-certifying id.
            var honest = new Description.Import(Encoding.UTF8.GetBytes(signerId), Description.FormatANPDescription, foreign, ops);
            byte[] signed = Description.SignImport(honest, Alg, seed);
            Description.ResolvedImport res = Description.VerifyImport(signed, Cose.PROFILE_PUBLIC, Alg, pk);
            Assert.Equal(signerId, res.AuthorityId);                         // the authority is the wrapping key, never the foreign identity
            Assert.Equal(Hex(honest.ForeignId()), Hex(res.ForeignId));       // it binds the exact foreign bytes
            // the foreign bytes DO embed a foreign identity (did:wba:...), which never becomes the authority.
            Assert.Contains("did:wba:foreign.example:agent", Encoding.UTF8.GetString(foreign));
            Assert.NotEqual("did:wba:foreign.example:agent", res.AuthorityId);

            // a signer that names a DIFFERENT importer than its own key id can only import AS ITSELF.
            var lying = new Description.Import(Encoding.UTF8.GetBytes("did:wba:foreign.example:agent"), Description.FormatANPDescription, foreign, ops);
            byte[] signedLie = Description.SignImport(lying, Alg, seed);
            var ex = Assert.Throws<NaalpException>(() => Description.VerifyImport(signedLie, Cose.PROFILE_PUBLIC, Alg, pk));
            Assert.Equal("ImporterMismatch", ex.Kind);
        }
    }
}

File: impl/kotlin/src/test/kotlin/DescriptionTests.kt
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
package sh.bubblefish.naalp

import java.io.File
import kotlin.system.exitProcess

//
// C18 signed description / directory / import known-answer test for the Kotlin SDK (design.md §21;
// R-DESC-1..8), graded against the shared independent corpus vectors/description/cases.json (NOT
// produced by this code).
//
// CORPUS-GRADED (pure, no signature): the Description body + head + content id and each operation body;
// the Directory body/head/id for both versions; fork detection at the FIRST-DIFFERING member position
// (a member-substitution fork at position 1, a truncated-member fork at position 1, a different version
// is a legitimate succession not a fork, a duplicate is no fork); the Import body/head/id and the bound
// foreign content id; a foreign format outside the closed set {1,2,3} rejected UnknownDescriptionFormat.
// ISOLATION (real FIPS-204 ML-DSA-65 via BouncyCastle — Kotlin is a full-signature port; the corpus
// carries NO signed vector, so these are demonstrated with a fixed local seed, stated honestly, NOT
// corpus-graded): a signed Description re-verifies to the identical operation table regardless of the
// serving host (offline / third-party re-serve); a DirectoryForkProof by one signer returns the
// first-differing position; and the confused-deputy rule — the attested importer MUST equal the
// verifying key's self-certifying signer id, else ImporterMismatch; a foreign identity never authorizes.
//
// KAT convention: a standalone main() exiting non-zero on any failure; the filename carries the "Tests"
// token the ten-language-parity gate indexes. Written test-first: [Desc] is absent until Description.kt
// lands, so this fails RED with a kotlinc "unresolved reference: Desc". The load-bearing mutation — the
// firstMemberDifference byte comparison made to never differ (if (!a[i].contentEquals(b[i])) ->
// if (false)) — flips "directory fork detected at first-differing position 1" on its assertion.
//

private var deFails = 0

private fun deCheck(name: String, got: String, want: String) {
    if (got == want) {
        println("  ok   $name")
    } else {
        deFails++
        println("  FAIL $name\n       got  $got\n       want $want")
    }
}

// ---- balanced-brace / regex JSON access (no JSON library on the Kotlin port) ----

private fun deFindVector(): File {
    var d: File? = File(".").absoluteFile
    repeat(6) {
        val cur = d ?: throw AssertionError("vectors/description/cases.json not found")
        val p = File(File(cur, "vectors"), "description/cases.json")
        if (p.isFile) return p
        d = cur.parentFile
    }
    throw AssertionError("vectors/description/cases.json not found from ${File(".").absolutePath}")
}

private fun deSection(scope: String, key: String): String {
    val m = Regex("\"" + Regex.escape(key) + "\"\\s*:\\s*").find(scope)
        ?: throw AssertionError("key not found: $key")
    var i = m.range.last + 1
    while (i < scope.length && scope[i].isWhitespace()) i++
    val open = scope[i]
    val close = when (open) { '{' -> '}'; '[' -> ']'; else -> throw AssertionError("not object/array at $key") }
    var depth = 0
    var inStr = false
    var esc = false
    val start = i
    while (i < scope.length) {
        val c = scope[i]
        if (inStr) {
            when {
                esc -> esc = false
                c == '\\' -> esc = true
                c == '"' -> inStr = false
            }
        } else {
            when (c) {
                '"' -> inStr = true
                open -> depth++
                close -> { depth--; if (depth == 0) return scope.substring(start, i + 1) }
            }
        }
        i++
    }
    throw AssertionError("unbalanced value at key: $key")
}

private fun deElements(arraySection: String): List<String> {
    val out = ArrayList<String>()
    var i = 1
    var inStr = false
    var esc = false
    var depth = 0
    var start = -1
    while (i < arraySection.length - 1) {
        val c = arraySection[i]
        if (inStr) {
            when {
                esc -> esc = false
                c == '\\' -> esc = true
                c == '"' -> inStr = false
            }
        } else {
            when (c) {
                '"' -> inStr = true
                '{' -> { if (depth == 0) start = i; depth++ }
                '}' -> { depth--; if (depth == 0 && start >= 0) { out.add(arraySection.substring(start, i + 1)); start = -1 } }
            }
        }
        i++
    }
    return out
}

// An array of quoted hex strings (member content-id lists), decoded to bytes.
private fun deHexArray(arraySection: String): List<ByteArray> {
    val out = ArrayList<ByteArray>()
    Regex("\"([0-9a-fA-F]*)\"").findAll(arraySection).forEach { out.add(Hex.decode(it.groupValues[1])) }
    return out
}

private fun deStr(scope: String, key: String): String {
    val m = Regex("\"" + Regex.escape(key) + "\"\\s*:\\s*\"([^\"]*)\"").find(scope)
        ?: throw AssertionError("string key not found: $key")
    return m.groupValues[1]
}

private fun deInt(scope: String, key: String): Long {
    val m = Regex("\"" + Regex.escape(key) + "\"\\s*:\\s*(\\d+)").find(scope)
        ?: throw AssertionError("int key not found: $key")
    return m.groupValues[1].toLong()
}

private fun deHex(s: String): ByteArray = Hex.decode(s)

private const val DE_ALG = Cose.ALG_MLDSA65
private val DE_SEED = ByteArray(32) { 5 }
private val DE_PK = Cose.mldsaKeygen("ML-DSA-65", DE_SEED)

private inline fun deKind(block: () -> Unit): String = try {
    block(); "no-error"
} catch (e: NaalpException) {
    e.kind
}

private fun deOp(block: String): Desc.Operation =
    Desc.Operation(deStr(block, "name"), deInt(block, "effect"), deInt(block, "requires_approval"))

private fun descriptionRun() {
    val json = deFindVector().readText(Charsets.UTF_8)

    // 1. Description: each operation body, the whole body/head/id, a parse round-trip, and per-operation
    //    effect + approval-declaration accessors.
    val descSec = deSection(json, "description")
    // The description's top-level body_hex sits AFTER the operations array (whose elements each also carry
    // a body_hex); scope the top-level read to the tail so the nested op body_hex cannot shadow it.
    val opsSection = deSection(descSec, "operations")
    val descTail = descSec.substring(descSec.indexOf(opsSection) + opsSection.length)
    val descBodyHex = deStr(descTail, "body_hex")
    val opElems = deElements(opsSection)
    val ops = ArrayList<Desc.Operation>()
    for (ob in opElems) {
        val op = deOp(ob)
        deCheck("description op ${op.name} body == oracle", Hex.encode(op.bytes()), deStr(ob, "body_hex"))
        ops.add(op)
    }
    val desc = Desc.Description(deHex(deStr(descSec, "service_hex")), ops)
    deCheck("description body == oracle", Hex.encode(desc.bytes()), descBodyHex)
    deCheck("description head == oracle", Hex.encode(desc.head()), deStr(descTail, "head_hex"))
    deCheck("description id == oracle", Hex.encode(desc.id()), deStr(descTail, "id_hex"))
    deCheck("description parse round-trip == oracle", Hex.encode(Desc.parseDescription(desc.bytes()).bytes()), descBodyHex)
    val purge = desc.operation("purge") ?: throw AssertionError("purge operation missing")
    deCheck("description op purge effect class == destructive", purge.effectClass().toString(), Policy.DESTRUCTIVE.toString())
    deCheck("description op purge requires approval", purge.requiresApprovalFlag().toString(), "true")
    val status = desc.operation("status") ?: throw AssertionError("status operation missing")
    deCheck("description op status requires no approval", status.requiresApprovalFlag().toString(), "false")

    // 2. Directory: both versions' body/head/id, and fork detection at the first-differing member
    //    position. THIS is the mutation's target assertion.
    val dirSec = deSection(json, "directory")
    val dirId = deHex(deStr(dirSec, "directory_hex"))
    val membersA = deHexArray(deSection(dirSec, "members_a_hex"))
    val dirA = Desc.Directory(dirId, deInt(dirSec, "version"), membersA)
    val aSec = deSection(dirSec, "a")
    deCheck("directory a body == oracle", Hex.encode(dirA.bytes()), deStr(aSec, "body_hex"))
    deCheck("directory a head == oracle", Hex.encode(dirA.head()), deStr(aSec, "head_hex"))
    deCheck("directory a id == oracle", Hex.encode(dirA.id()), deStr(aSec, "id_hex"))

    val forkSec = deSection(dirSec, "fork")
    val membersB = deHexArray(deSection(forkSec, "members_b_hex"))
    val dirB = Desc.Directory(dirId, deInt(dirSec, "version"), membersB)
    val bSec = deSection(forkSec, "b")
    deCheck("directory fork b body == oracle", Hex.encode(dirB.bytes()), deStr(bSec, "body_hex"))
    deCheck("directory fork b head == oracle", Hex.encode(dirB.head()), deStr(bSec, "head_hex"))
    deCheck("directory fork b id == oracle", Hex.encode(dirB.id()), deStr(bSec, "id_hex"))
    val (fpos, ffork) = Desc.detectFork(dirA, dirB)
    deCheck("directory fork detected", ffork.toString(), "true")
    deCheck("directory fork detected at first-differing position 1", fpos.toString(), deInt(forkSec, "first_differing_position").toString())

    // a truncated member list forks at the length of the shorter list.
    val lenSec = deSection(dirSec, "length_fork")
    val membersShort = deHexArray(deSection(lenSec, "members_short_hex"))
    val dirShort = Desc.Directory(dirId, deInt(dirSec, "version"), membersShort)
    val (lpos, lfork) = Desc.detectFork(dirA, dirShort)
    deCheck("directory length fork detected", lfork.toString(), "true")
    deCheck("directory length fork at position 1", lpos.toString(), deInt(lenSec, "first_differing_position").toString())

    // a different version is a legitimate succession, not a fork.
    val dvSec = deSection(dirSec, "different_version")
    val dirV8 = Desc.Directory(dirId, deInt(dvSec, "version"), membersB)
    val (_, dvFork) = Desc.detectFork(dirA, dirV8)
    deCheck("directory different version is not a fork", dvFork.toString(), "false")

    // a duplicate (identical members) is not a fork.
    val (_, dupFork) = Desc.detectFork(dirA, dirA)
    deCheck("directory duplicate is not a fork", dupFork.toString(), "false")

    // 3. Import: body/head/id, the bound foreign content id, a parse round-trip, and an unknown foreign
    //    format rejected on decode.
    val impSec = deSection(json, "import")
    val impOps = deElements(deSection(impSec, "operations")).map { deOp(it) }
    val imp = Desc.Import(deHex(deStr(impSec, "importer_hex")), deInt(impSec, "format"), deHex(deStr(impSec, "foreign_hex")), impOps)
    deCheck("import body == oracle", Hex.encode(imp.bytes()), deStr(impSec, "body_hex"))
    deCheck("import head == oracle", Hex.encode(imp.head()), deStr(impSec, "head_hex"))
    deCheck("import id == oracle", Hex.encode(imp.id()), deStr(impSec, "id_hex"))
    deCheck("import foreign id == oracle", Hex.encode(imp.foreignId()), deStr(impSec, "foreign_id_hex"))
    deCheck("import parse round-trip == oracle", Hex.encode(Desc.parseImport(imp.bytes()).bytes()), deStr(impSec, "body_hex"))
    val unSec = deSection(impSec, "unknown_format")
    deCheck("import unknown format rejected", deKind { Desc.parseImport(deHex(deStr(unSec, "body_hex"))) }, deStr(unSec, "reject"))

    // 4. ISOLATION (real ML-DSA): the signed paths.
    // 4a. offline / third-party re-serve: a signed Description re-verifies to the identical bytes.
    val signedDesc = Desc.signDescription(desc, DE_ALG, DE_SEED)
    deCheck("iso signed Description re-verifies to identical bytes", Hex.encode(Desc.verifyDescription(signedDesc, Cose.PROFILE_PUBLIC, DE_ALG, DE_PK).bytes()), descBodyHex)
    // a tampered signature is BadSignature (fail-closed).
    val tampered = signedDesc.copyOf(); tampered[tampered.size - 1] = (tampered[tampered.size - 1].toInt() xor 1).toByte()
    deCheck("iso tampered Description rejected BadSignature", deKind { Desc.verifyDescription(tampered, Cose.PROFILE_PUBLIC, DE_ALG, DE_PK) }, "BadSignature")

    // 4b. DirectoryForkProof by one signer returns the first-differing position; a non-fork is rejected.
    val signedA = Desc.signDirectory(dirA, DE_ALG, DE_SEED)
    val signedB = Desc.signDirectory(dirB, DE_ALG, DE_SEED)
    val proof = Desc.DirectoryForkProof("accused".toByteArray(), signedA, signedB)
    deCheck("iso DirectoryForkProof position == 1", proof.verify(Cose.PROFILE_PUBLIC, DE_ALG, DE_PK).toString(), deInt(forkSec, "first_differing_position").toString())
    val signedV8 = Desc.signDirectory(dirV8, DE_ALG, DE_SEED)
    val notFork = Desc.DirectoryForkProof("accused".toByteArray(), signedA, signedV8)
    deCheck("iso non-fork proof rejected DirForkProofInvalid", deKind { notFork.verify(Cose.PROFILE_PUBLIC, DE_ALG, DE_PK) }, "DirForkProofInvalid")

    // 4c. confused-deputy: the attested importer MUST equal the verifying key's signer id.
    val keyId = Identity.signerId(DE_ALG, DE_PK)
    val goodImport = Desc.Import(keyId.toByteArray(Charsets.UTF_8), Desc.FORMAT_ANP_DESCRIPTION, deHex(deStr(impSec, "foreign_hex")), impOps)
    val signedImport = Desc.signImport(goodImport, DE_ALG, DE_SEED)
    val resolved = Desc.verifyImport(signedImport, Cose.PROFILE_PUBLIC, DE_ALG, DE_PK)
    deCheck("iso verifyImport authority == signer id", resolved.authorityId, keyId)
    deCheck("iso verifyImport foreign id bound == oracle", Hex.encode(resolved.foreignId), deStr(impSec, "foreign_id_hex"))
    // a foreign importer that is not the verifying key's signer id never authorizes.
    val mismatchImport = Desc.Import(deHex(deStr(impSec, "importer_hex")), Desc.FORMAT_ANP_DESCRIPTION, deHex(deStr(impSec, "foreign_hex")), impOps)
    val signedMismatch = Desc.signImport(mismatchImport, DE_ALG, DE_SEED)
    deCheck("iso verifyImport foreign importer rejected ImporterMismatch", deKind { Desc.verifyImport(signedMismatch, Cose.PROFILE_PUBLIC, DE_ALG, DE_PK) }, "ImporterMismatch")
}

fun main() {
    println("description conformance (Kotlin) — graded vs vectors/description/cases.json")
    descriptionRun()
    println(if (deFails == 0) "DescriptionTests: PASS" else "DescriptionTests: FAIL ($deFails)")
    exitProcess(if (deFails == 0) 0 else 1)
}
